### PR TITLE
ESC-236 date validation rules and fixes

### PR DESCRIPTION
**Pull request by carwynellis**
Summary of changes
* refactored date form validation introducing a `FormProvider` trait and moving the date form into a new `forms` package - other forms can be moved here later with unit tests introduced for each
* added tests for the date form validation
* fixed date form validation bugs
  * entering today's date triggered the 'date in the future' validation error
  * entering an invalid date e.g. `99/99/9999` threw an exception
* added new validation rule - earliest claim date must be within the current or previous 2 tax years
* refactoring
  * fixed all compiler warnings
  * introduced common handling for missing data on session - need to determine the best course of action here (e.g. should we just redirect to a session expired page and force the user to login again?) - left as a TODO for now
  * date validation has been refactored and simplified where possible
  * case classes have been removed from the controllers package object and moved into models
* added the `scoverage` plugin but no minimum coverage rule has been configured yet

Jira ticket: https://jira.tools.tax.service.gov.uk/browse/ESC-236

### CODE REVIEW

File: test/uk/gov/hmrc/eusubsidycompliancefrontend/forms/ClaimDateFormProviderSpec.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+import java.time.{LocalDate, ZoneId}
+
+class ClaimDateFormProviderSpec extends AnyWordSpecLike with Matchers {
+
+  private val day = 1
+  private val month = 1
+  private val year = 2022
+
+  private val fakeTimeProvider: TimeProvider = new TimeProvider {
```
Comment (priyankadhiman3006) [inline test/uk/gov/hmrc/eusubsidycompliancefrontend/forms/ClaimDateFormProviderSpec.scala:33]: You can make it def and we can test it using various dates? like 1 jan , 4 april, 5 april. what do you think?

Comment (carwynellis) [reply test/uk/gov/hmrc/eusubsidycompliancefrontend/forms/ClaimDateFormProviderSpec.scala:33]: Yeah I can pull this out so other tests can use it. Will do that now.

Comment (carwynellis) [reply test/uk/gov/hmrc/eusubsidycompliancefrontend/forms/ClaimDateFormProviderSpec.scala]: Pushed an update for that now.

File: app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/BaseController.scala [hunk 2]
```diff
@@ -35,6 +35,12 @@ class BaseController @Inject()(mcc: MessagesControllerComponents) extends Fronte
     text.transform[String](_.trim, s => s).verifying(combine(required(key), constraint(key, regex)))
   }
 
+  // TODO - if data is missing it's likely the session has expired.
+  //        Redirect to a session expired page instead?
+  protected def handleMissingSessionData(dataLabel: String) =
+    throw new IllegalStateException(s"$dataLabel data not present on session")
```
Comment (priyankadhiman3006) [inline app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/BaseController.scala:41]: minor: In session

**Review by priyankadhiman3006 — APPROVED** [review]

File: app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/SubsidyController.scala [hunk 3]
```diff
@@ -187,6 +187,7 @@ class SubsidyController @Inject()(
               Ok(addClaimEoriPage(claimEoriForm.fill(OptionalEORI(a,x)), journey.previous))
             )
           }
+      case _ => handleMissingSessionData("Subsidy journey")
```
Comment (sam-miele) [inline app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/SubsidyController.scala:190]: NABD a lot of repeating here.

File: app/uk/gov/hmrc/eusubsidycompliancefrontend/services/BusinessEntityJourney.scala [hunk 4]
```diff
@@ -53,6 +54,8 @@ object BusinessEntityJourney {
           empty.eori.copy(value = Some(eori)),
           empty.contact.copy(value = cd)
         )
+      // TODO - what is the correct behaviour here?
+      case None => BusinessEntityJourney()
```
Comment (sam-miele) [inline app/uk/gov/hmrc/eusubsidycompliancefrontend/services/BusinessEntityJourney.scala:58]: We should probably think about this. I think we should redirect at first thought. But we'll need to put this in a lot of places.

File: app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/SubsidyController.scala (hunk 3 above)
Comment (carwynellis) [reply app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/SubsidyController.scala:190]: Yeah it's to fix the warning about the `None` case not being matched in all these places. Needs another pass really to see if there's a better way to structure this, but this was a first step to just fix the warnings for now.

File: app/uk/gov/hmrc/eusubsidycompliancefrontend/services/BusinessEntityJourney.scala (hunk 4 above)
Comment (carwynellis) [reply app/uk/gov/hmrc/eusubsidycompliancefrontend/services/BusinessEntityJourney.scala:58]: OK - I'll leave the TODO here as a reminder.

**Review by sam-miele — APPROVED** [review]

**Pull request merged**